controllers/medusa/medusarestorejob_controller.go

```diff
@@ -0,0 +1,310 @@
+			}
+			req.Log.Error(err, "Failed to patch the CassandraDatacenter")
+			return err
+		}
+	}
+
+	if req.MedusaRestoreModified() {
+		if err := r.Status().Patch(ctx, req.RestoreJob, req.GetRestorePatch()); err != nil {
+			req.Log.Error(err, "Failed to patch the MedusaRestoreJob")
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (r *MedusaRestoreJobReconciler) applyUpdatesAndRequeue(ctx context.Context, req *medusa.RestoreRequest) (ctrl.Result, error) {
+	if err := r.applyUpdates(ctx, req); err != nil {
+		req.Log.Error(err, "Failed to apply updates")
+		return ctrl.Result{RequeueAfter: r.DefaultDelay}, err
+	}
+	return ctrl.Result{RequeueAfter: r.DefaultDelay}, nil
+}
+
+// podTemplateSpecUpdateComplete checks that the pod template spec changes, namely the ones
+// with the restore container, have been pushed down to the StatefulSets. Return true if
+// the changes have been applied.
+func (r *MedusaRestoreJobReconciler) podTemplateSpecUpdateComplete(ctx context.Context, req *medusa.RestoreRequest) (bool, error) {
+	if updated := cassandra.DatacenterUpdatedAfter(req.RestoreJob.Status.DatacenterStopped.Time, req.Datacenter); !updated {
+		return false, nil
+	}
+
+	// It may not be sufficient to check for the update only via status conditions. We will
+	// check the template spec of the StatefulSets to be certain that the update has been
+	// applied. We do this in order to avoid an extra rolling restart after the
+	// StatefulSets are scaled back up.
+
+	statefulsetList := &appsv1.StatefulSetList{}
+	labels := client.MatchingLabels{cassdcapi.ClusterLabel: req.Datacenter.Spec.ClusterName, cassdcapi.DatacenterLabel: req.Datacenter.Name}
+
+	if err := r.List(ctx, statefulsetList, labels); err != nil {
+		req.Log.Error(err, "Failed to get StatefulSets")
+		return false, err
+	}
+
+	for _, statefulset := range statefulsetList.Items {
+		container := getRestoreInitContainerFromStatefulSet(&statefulset)
+
+		if container == nil {
+			return false, nil
+		}
+
+		if !containerHasEnvVar(container, backupNameEnvVar, req.MedusaBackup.ObjectMeta.Name) {
+			return false, nil
+		}
+
+		if !containerHasEnvVar(container, restoreKeyEnvVar, req.RestoreJob.Status.RestoreKey) {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+func (r *MedusaRestoreJobReconciler) prepareRestore(ctx context.Context, request *medusa.RestoreRequest) (bool, error) {
+	// Create a prepare_restore medusa task to create the mapping files in each pod.
+	// Returns true if the reconcile needs to be requeued, false otherwise.
+	prepare := &medusav1alpha1.MedusaTask{}
+	if err := r.Client.Get(context.Background(), types.NamespacedName{Name: request.RestoreJob.Status.RestoreKey, Namespace: request.RestoreJob.Namespace}, prepare); err != nil {
+		if errors.IsNotFound(err) {
+			// Create the sync task
+			prepare = &medusav1alpha1.MedusaTask{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      request.RestoreJob.Status.RestoreKey,
```

Comment:
Can we change the task name to something more descriptive, maybe something like `<cluster-name>-<dc-name>-prepare-restore`?  Seeing task objects with UUIDs for their names isn't very helpful. Alternatively we could update the CRD to include the operation type and cassdc name in the output for `kubectl get`.

Reply:
to be fair, it'll be created in the namespace of the cluster that's being restored.
If multiple restores are done, we need a name that distinguish them, and using the restore key seemed like a good idea.
I'm afraid that we may create names that are too long if we stuff in things like the cluster name 😕 

> Alternatively we could update the CRD to include the operation type and cassdc name in the output for kubectl get.

Would you add that in the status of the CRD?

Reply:
>I'm afraid that we may create names that are too long if we stuff in things like the cluster name

Definitely a valid concern.

>Would you add that in the status of the CRD?

Not in the status. There is a `kubebuilder:printcolumn:name` annotation. See https://github.com/k8ssandra/k8ssandra-operator/blob/main/apis/stargate/v1alpha1/stargate_types.go#L283 for examples. Would make for a nice follow up enhancement. I don't think it's needed for this PR.